### PR TITLE
Switch to Arcade master branch when promoting a build that uses Arcade 3.x

### PR DESCRIPTION
**Pull request by JohnTortugo**
Relates to: https://github.com/dotnet/arcade/issues/5408

- Check if the user specified "release/3.x" as --source-branch and if so warns and switch to "master"
- Check if promoting a build that uses an Arcade 3.x and if so warns and switch to "master"

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/AddBuildToChannelOperation.cs [hunk 1]
```diff
@@ -316,6 +350,13 @@ private async Task<(string sourceBranch, string sourceVersion)> GetSourceBranchI
             if (hasSourceBranch)
             {
                 _options.SourceBranch = GitHelpers.NormalizeBranchName(_options.SourceBranch);
+
+                if (_options.SourceBranch.EndsWith("release/3.x", StringComparison.OrdinalIgnoreCase))
+                {
+                    Console.WriteLine($"Warning: Arcade branch {_options.SourceBranch} doesn't support build promotion. Please try specifiying --source-branch 'master'.");
+                    Console.WriteLine("Switching source branch to Arcade master.");
```
Comment (riarenas) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/AddBuildToChannelOperation.cs:357]: So If I understand this correctly, we tell the user to consider switching to using master, but we switch for them anyways?

Comment (JohnTortugo) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/AddBuildToChannelOperation.cs:357]: Yes, that's correct.

Comment (riarenas) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/AddBuildToChannelOperation.cs:357]: We should probably just go ahead and inform them that we're doing it. I don't see much value in telling them to do something that is getting handled automatically.

**Comment by JohnTortugo** [comment]
> Could we do this differently? Couldn't we simply identify what version of arcade introduced the promotion pipeline (it's a 5.0 version) and then compare against the version that the user has in their version file. If >, then good to use promotion. Otherwise switch to master?

Perhaps I'm missing something but I think that's what we currently do here. [See the check on line 419 for instance.](https://github.com/dotnet/arcade-services/pull/1150/files#diff-aa5a491937907bb6146e94e0ab683fd3R419) The additional validations here are because:

- The user can try and override the default branch and we shouldn't allow 'release/3.x'
- Possibly, we can have a repo using a 3.x SDK that was created after we introduced the promotion pipeline
- Even using a 5.x SDK the user can try to promote to a channel that doesn't have YAML in the master branch

**Comment by mmitche** [comment]
> > Could we do this differently? Couldn't we simply identify what version of arcade introduced the promotion pipeline (it's a 5.0 version) and then compare against the version that the user has in their version file. If >, then good to use promotion. Otherwise switch to master?
> 
> Perhaps I'm missing something but I think that's what we currently do here. [See the check on line 419 for instance.](https://github.com/dotnet/arcade-services/pull/1150/files#diff-aa5a491937907bb6146e94e0ab683fd3R419) The additional validations here are because:
> 
> * The user can try and override the default branch and we shouldn't allow 'release/3.x'
> * Possibly, we can have a repo using a 3.x SDK that was created after we introduced the promotion pipeline
> * Even using a 5.x SDK the user can try to promote to a channel that doesn't have YAML in the master branch

I see. It's similar, in that this would cover the cases where someone is on an old 5.x that doesn't have  promotion capabilities. Anyways, it's probably fine as is.

**Pull request merged**